lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,287 @@
+
+
+class BlockRecordSet(frozenset):
+    """
+    An immutable ordered collection of BlockRecord objects.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(BlockRecordSet, self).__init__(*args, **kwargs)
+        self._json = None
+        self._hash = None
+
+    def to_json(self):
+        """
+        Return a JSON-serialized version of the list of block records, using a
+        stable ordering.
+        """
+        if self._json is None:
+            sorted_blocks = sorted(self, key=attrgetter('locator'))
+            # Remove spaces from separators for more compact representation
+            self._json = json.dumps(
+                [block._asdict() for block in sorted_blocks],
+                separators=(',', ':'),
+                sort_keys=True,
+            )
+        return self._json
+
+    @classmethod
+    def from_json(cls, blockrecord_json):
+        """
+        Return a BlockRecordSet from a json list.
+        """
+        block_dicts = json.loads(blockrecord_json)
+        record_generator = (BlockRecord(**block) for block in block_dicts)
+        return cls(record_generator)
+
+    def to_hash(self):
+        """
+        Return a hashed version of the list of block records.
+
+        This currently hashes using sha1, and returns a base64 encoded version
+        of the binary digest.  In the future, different algorithms could be
+        supported by adding a label indicated which algorithm was used, e.g.,
+        "sha256$j0NDRmSPa5bfid2pAcUXaxCm2Dlh3TwayItZstwyeqQ=".
+        """
+        if self._hash is None:
+            self._hash = b64encode(sha1(self.to_json()).digest())
+        return self._hash
+
+
+class VisibleBlocksQuerySet(models.QuerySet):
+    """
+    A custom QuerySet representing VisibleBlocks.
+    """
+
+    def create_from_blockrecords(self, blocks):
+        """
+        Creates a new VisibleBlocks model object.
+
+        Argument 'blocks' should be a BlockRecordSet.
+        """
+
+        if not isinstance(blocks, BlockRecordSet):
+            blocks = BlockRecordSet(blocks)
+
+        model, _ = self.get_or_create(hashed=blocks.to_hash(), defaults={'blocks_json': blocks.to_json()})
+        return model
+
+    def hash_from_blockrecords(self, blocks):
+        """
+        Return the hash for a given BlockRecordSet, serializing the records if
+        possible, but returning the hash even if an IntegrityError occurs.
+        """
+
+        if not isinstance(blocks, BlockRecordSet):
+            blocks = BlockRecordSet(blocks)
+
+        try:
+            model = self.create_from_blockrecords(blocks)
+        except IntegrityError:
+            # If an integrity error occurs, the VisibleBlocks model we want to
+            # create already exists.  The hash is still the correct value.
+            return blocks.to_hash()
+        else:
+            # No error occurred
+            return model.hashed
```

Comment:
@nasthagiri I couldn't think of a better refactoring than this.  Now that I'm looking at it, it doesn't look terrible. :shrugs:.  I also used this opportunity to move the HashRecordSet coercion further in.  Now it only occurs inside VisibleBlocksQuerySets.